CI improvements

kuznetsss opened the pull request:
Fixes #865, #869

Improvements:
- move duplicated code into actions
- use ccache to speed up compilation
- isolate build on mac (don't use home directory)
- generate conan profile on each run
- cache `.conan/data` 
- cache `.ccache`

Doubts:
- there is still duplicated code on the top level. I tried to create an action from it, but log doesn't look nice after that
- ~~key for conan data cache is hash from conan file. It may bring troubles because it is using only the dependencies from the first level~~
- this PR overrides docker image for linux. But conan's docker image is build on ubuntu 20.04 while packages binaries from artifactory are built using ubuntu 22.04.

kuznetsss committed:
Generate conan profile in CI

kuznetsss committed:
Move linux build into main workflow

kuznetsss committed:
Add saving/restoring conan data

godexsoft commented on .github/workflows/build.yml:
```diff
@@ -18,41 +18,54 @@ jobs:
+      - name: Restore conan cache
+        uses: actions/cache/restore@v3
+        with:
+          path: path/to/dependencies
```
Looks like some sort of placeholder path?

kuznetsss replied:
Fixed, thanks

kuznetsss committed:
Move cache to Linux

kuznetsss committed:
Fix error

kuznetsss committed:
Change key to hash from conanfile

kuznetsss committed:
Fix path error

kuznetsss committed:
Populate cache only in develop branch

kuznetsss committed:
Big refactor

- Move duplicated code to actions
- Isolate mac build from home directory
- Separate ccache and conan caches

kuznetsss committed:
Fix errors

kuznetsss committed:
Change ccache cache name and fix errors

kuznetsss committed:
Always populate cache

kuznetsss committed:
Use newer ccache on Linux

kuznetsss committed:
Strip tests

cindyyan317 commented on .github/actions/save_cache/action.yml:
```diff
@@ -0,0 +1,34 @@
+      uses: actions/cache/save@v3
+      with:
+        path: ${{ inputs.conan_dir }}/data
+        key: clio-conan_data-${{ runner.os }}-develop-${{ hashFiles('./conanfile.py') }}
```
i am thinking if there is any chance , the runner.os does not change and conanfile.py does not change, but the cache is outdated? for example, the compiler gets updated? 
Do we need to also consider to add build mode (debug/release)?

kuznetsss replied:
We can always manually delete cache here: https://github.com/XRPLF/clio/actions/caches
We have only release builds now, we could change key later if needed later. Cache miss costs only about 1 hour of CI work

legleux replied:
The Conan packages' hashes encapsulate the compiler version so that should invalidate the cache right?

kuznetsss replied:
If we update compiler, conan will not use packages in local cache. But with the current implementation because `conanfile.py` didn't change, CI will not upload a new cache to the github storage.

I found a better solution:
```bash
conan info . -j info.json
cat info.json | jq '.[] | "\(.id), \(.package_revision)"' | shasum -a 256 -
```
This will take id and revision from every conan package used and calculate sha256 from it. So if we change compiler packages revisions will be different and the sha256 hash will be different.

kuznetsss committed:
Better conan hash

godexsoft reviewed (approved):
LGTM 👍

kuznetsss committed:
Merge branch 'develop' into Improve_CI
